[Docs] Adding Store API Docs

osalpekar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45543 [Docs] Adding Store API Docs**

This PR adds documentation for the c10d Store to the public docs. Previously these docs were missing altogether even though we exposed a Python API for our distributed key-value store. This Store API could be useful for users looking for a quick way to share information across workers when using PyTorch distributed.

See example render below:
<img width="837" alt="Screen Shot 2020-09-29 at 11 42 33 PM" src="https://user-images.githubusercontent.com/14008860/94864458-dc6b8380-03f0-11eb-9776-81a45f93f888.png">

Differential Revision: [D24005598](https://our.internmc.facebook.com/intern/diff/D24005598/)

osalpekar committed:
[Docs] Adding Store API Docs

This PR adds documentation for the c10d Store to the public docs. Previously these docs were missing although we exposed a lightly-used (but potentially useful) Python API for our distributed key-value store.

Differential Revision: [D24005598](https://our.internmc.facebook.com/intern/diff/D24005598/)

[ghstack-poisoned]

osalpekar committed:
Update on "[Docs] Adding Store API Docs"


This PR adds documentation for the c10d Store to the public docs. Previously these docs were missing altogether even though we exposed a Python API for our distributed key-value store. This Store API could be useful for users looking for a quick way to share information across workers when using PyTorch distributed.

Differential Revision: [D24005598](https://our.internmc.facebook.com/intern/diff/D24005598/)

[ghstack-poisoned]

osalpekar committed:
Update on "[Docs] Adding Store API Docs"


This PR adds documentation for the c10d Store to the public docs. Previously these docs were missing altogether even though we exposed a Python API for our distributed key-value store. This Store API could be useful for users looking for a quick way to share information across workers when using PyTorch distributed.

Differential Revision: [D24005598](https://our.internmc.facebook.com/intern/diff/D24005598/)

[ghstack-poisoned]

osalpekar committed:
Update on "[Docs] Adding Store API Docs"


This PR adds documentation for the c10d Store to the public docs. Previously these docs were missing altogether even though we exposed a Python API for our distributed key-value store. This Store API could be useful for users looking for a quick way to share information across workers when using PyTorch distributed.

Differential Revision: [D24005598](https://our.internmc.facebook.com/intern/diff/D24005598/)

[ghstack-poisoned]

osalpekar committed:
Update on "[Docs] Adding Store API Docs"


This PR adds documentation for the c10d Store to the public docs. Previously these docs were missing altogether even though we exposed a Python API for our distributed key-value store. This Store API could be useful for users looking for a quick way to share information across workers when using PyTorch distributed.

Differential Revision: [D24005598](https://our.internmc.facebook.com/intern/diff/D24005598/)

[ghstack-poisoned]

osalpekar committed:
Update on "[Docs] Adding Store API Docs"


This PR adds documentation for the c10d Store to the public docs. Previously these docs were missing altogether even though we exposed a Python API for our distributed key-value store. This Store API could be useful for users looking for a quick way to share information across workers when using PyTorch distributed.

Differential Revision: [D24005598](https://our.internmc.facebook.com/intern/diff/D24005598/)

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -306,46 +324,232 @@ They are used in specifying strategies for reduction collectives, e.g.,
+
+Example::
+    >>> import torch.distributed as dist
+    >>> # Using TCPStore as an example, other store types can also be used
```
nit: Other store types won't work with this example due to delete_key exceptions right?

osalpekar replied:
Right, good catch

osalpekar committed:
Update on "[Docs] Adding Store API Docs"


This PR adds documentation for the c10d Store to the public docs. Previously these docs were missing altogether even though we exposed a Python API for our distributed key-value store. This Store API could be useful for users looking for a quick way to share information across workers when using PyTorch distributed.

Differential Revision: [D24005598](https://our.internmc.facebook.com/intern/diff/D24005598/)

[ghstack-poisoned]

osalpekar committed:
Update on "[Docs] Adding Store API Docs"


This PR adds documentation for the c10d Store to the public docs. Previously these docs were missing altogether even though we exposed a Python API for our distributed key-value store. This Store API could be useful for users looking for a quick way to share information across workers when using PyTorch distributed.

Differential Revision: [D24005598](https://our.internmc.facebook.com/intern/diff/D24005598/)

[ghstack-poisoned]

pritamdamania87 commented on docs/source/distributed.rst:
```diff
@@ -270,6 +270,30 @@ The machine with rank 0 will be used to set up all connections.
+Key-Value Stores: :class:`~torch.distributed.TCPStore`,
+:class:`~torch.distributed.FileStore`, and :class:`~torch.distributed.HashStore`.
+
+.. autoclass:: TCPStore
```
Should we also add Store here? Similar to torch.optim where there is documentation about the base class.

pritamdamania87 commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -306,46 +324,232 @@ They are used in specifying strategies for reduction collectives, e.g.,
-              py::call_guard<py::gil_scoped_release>())
+              py::call_guard<py::gil_scoped_release>(),
+              R"(
+add(key: str, amount: int) -> None
+
+The first call to add for a given ``key`` creates a counter associated
+with ``key`` in the store, initialized to ``amount``. Subsequent calls to add
+with the same ``key`` increment the counter by the specified ``amount``.
```
What is the behavior if the key is already present via `set` call? We should document that as well.

osalpekar replied:
It seems to throw an exception - will make a note about it in the docs.

pritamdamania87 commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -306,46 +324,232 @@ They are used in specifying strategies for reduction collectives, e.g.,
+num_keys() -> int
+
+Returns the number of keys set in the store. Note that this number will typically
+be one greater than the number of keys added by :meth:`~torch.distributed.store.get`
```
This should be `set` and not `get`?

pritamdamania87 commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -306,46 +324,232 @@ They are used in specifying strategies for reduction collectives, e.g.,
-  shared_ptr_class_<::c10d::HashStore>(module, "HashStore", store)
+  shared_ptr_class_<::c10d::HashStore>(module, "HashStore", store,
+      R"(
+A thread-safe store implementation based on an underlying hashmap (C++ std::unordered_map).
```
Don't add implementation details like c++ unordered_map. Also, mention that this store can only be used within the same process and not across processes.

osalpekar committed:
Update on "[Docs] Adding Store API Docs"


This PR adds documentation for the c10d Store to the public docs. Previously these docs were missing altogether even though we exposed a Python API for our distributed key-value store. This Store API could be useful for users looking for a quick way to share information across workers when using PyTorch distributed.

Differential Revision: [D24005598](https://our.internmc.facebook.com/intern/diff/D24005598/)

[ghstack-poisoned]

osalpekar committed:
Update on "[Docs] Adding Store API Docs"


This PR adds documentation for the c10d Store to the public docs. Previously these docs were missing altogether even though we exposed a Python API for our distributed key-value store. This Store API could be useful for users looking for a quick way to share information across workers when using PyTorch distributed.

Differential Revision: [D24005598](https://our.internmc.facebook.com/intern/diff/D24005598/)

[ghstack-poisoned]

osalpekar committed:
Update on "[Docs] Adding Store API Docs"


This PR adds documentation for the c10d Store to the public docs. Previously these docs were missing altogether even though we exposed a Python API for our distributed key-value store. This Store API could be useful for users looking for a quick way to share information across workers when using PyTorch distributed.

Differential Revision: [D24005598](https://our.internmc.facebook.com/intern/diff/D24005598/)

[ghstack-poisoned]

osalpekar committed:
Update on "[Docs] Adding Store API Docs"


This PR adds documentation for the c10d Store to the public docs. Previously these docs were missing altogether even though we exposed a Python API for our distributed key-value store. This Store API could be useful for users looking for a quick way to share information across workers when using PyTorch distributed.

Differential Revision: [D24005598](https://our.internmc.facebook.com/intern/diff/D24005598/)

[ghstack-poisoned]

osalpekar committed:
Update on "[Docs] Adding Store API Docs"


This PR adds documentation for the c10d Store to the public docs. Previously these docs were missing altogether even though we exposed a Python API for our distributed key-value store. This Store API could be useful for users looking for a quick way to share information across workers when using PyTorch distributed.

Differential Revision: [D24005598](https://our.internmc.facebook.com/intern/diff/D24005598/)

[ghstack-poisoned]

osalpekar committed:
Update on "[Docs] Adding Store API Docs"


This PR adds documentation for the c10d Store to the public docs. Previously these docs were missing altogether even though we exposed a Python API for our distributed key-value store. This Store API could be useful for users looking for a quick way to share information across workers when using PyTorch distributed.

See example render below:
<img width="837" alt="Screen Shot 2020-09-29 at 11 42 33 PM" src="https://user-images.githubusercontent.com/14008860/94864458-dc6b8380-03f0-11eb-9776-81a45f93f888.png">

Differential Revision: [D24005598](https://our.internmc.facebook.com/intern/diff/D24005598/)

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -284,7 +284,11 @@ They are used in specifying strategies for reduction collectives, e.g.,
-          module, "Store")
+          module, "Store",
+          R"(
+Base class for the 3 store implementations: (:class:`~torch.distributed.TCPStore`,
```
Probably just mention this is the base class for all store implementations. There could be other user implementations on top of this apart from these 3.

osalpekar committed:
Update on "[Docs] Adding Store API Docs"


This PR adds documentation for the c10d Store to the public docs. Previously these docs were missing altogether even though we exposed a Python API for our distributed key-value store. This Store API could be useful for users looking for a quick way to share information across workers when using PyTorch distributed.

See example render below:
<img width="837" alt="Screen Shot 2020-09-29 at 11 42 33 PM" src="https://user-images.githubusercontent.com/14008860/94864458-dc6b8380-03f0-11eb-9776-81a45f93f888.png">

Differential Revision: [D24005598](https://our.internmc.facebook.com/intern/diff/D24005598/)

[ghstack-poisoned]
